tests/performance/config/perf-driver/policies/test-continuous-n-apps.yml

```diff
@@ -66,6 +66,7 @@ channels:
     deploy:
       - type: app
         repeat: "{{apps}}"
+        parallel: 100
```

Comment:
Oh yes. It's was quite complicated, especially since we also need a retry and a back-off behaviour in _addition_ to the parallel requests.

Everything is in this stand-alone module : https://github.com/mesosphere/dcos-perf-test-driver/blob/master/performance/driver/classes/channel/utils/bulk.py

Reply:
Just out of curiosity, why did you not use Python's [multiprocessing module](https://docs.python.org/3/library/multiprocessing.html) instead of a custom executor service? Multi processing is incredibly hard in Python. It might even make sense to start a [drill](https://github.com/fcsonline/drill) benchmark from Python.

Reply:
Generally I avoid multiprocessing when I don't have CPU-bound problems. Mainly because I instinctively think that in such cases the IPC cost is bigger than the operation itself. In particularly, the driver internally dispatches thousands of messages per second, and broadcasting them to multiple processes will have some decent impact. 

In our case, the problem is I/O bound so we won't really benefit from multiprocessing. And since we are not stress-testing the endpoint, but we are actually waiting for a marathon operation to complete (which is much longer than the actual HTTP operation), we are more than safe with the current implementation.

As a matter of fact, since you mentioned this earlier, I have tried to switch the core thread pool (implemented in the [Event Bus](https://github.com/mesosphere/dcos-perf-test-driver/blob/master/performance/driver/core/eventbus.py#L26) ), from thread-based to process-based, since the `threading` and the `multiprocessing` have an identical API interface, and I did not observe any significant difference on the performance nor on the results... but we could try again if you like.

Reply:
> In our case, the problem is I/O bound so we won't really benefit from multiprocessing. 

I fully agree but it seems you wrote a threadpool there. I did not go through the complete code but it seems that either the multiprocessing module or asyncio would be safer choices since they've been tested and used by a larger crowd.

Reply:
That's true. But again, as long as you are doing I/O-bound operations on the threads, the GIL is the least of your concerns. Your python threads will behave almost identical to system threads. (Ps. `asyncio` is also affected by GIL, though it's not affected by context-switching, which makes it a teensy bit better on the performance).

However, keep in mind that the driver evolved into it's current state, it was not designed like so. In the beginning it was a pure, single-threaded application, and the multi-threading (and eventually the thread-pool) was introduced as a workaround to allow multiple blocking operations to happen at the same time.

If I was to start again from scratch, I am definitely going to use `asyncio`. So if you are interested on a re-write, let's pair on it 😄